src/analyze/analyze.c

```diff
@@ -192,22 +192,27 @@ static int bus_get_unit_property_strv(sd_bus *bus, const char *path, const char
 }
 
 static int compare_unit_time(const void *a, const void *b) {
-        return compare(((struct unit_times *)b)->time,
-                       ((struct unit_times *)a)->time);
+        return compare((*(struct unit_times **)b)->time,
+                       (*(struct unit_times **)a)->time);
 }
 
 static int compare_unit_start(const void *a, const void *b) {
-        return compare(((struct unit_times *)a)->activating,
-                       ((struct unit_times *)b)->activating);
+        return compare((*(struct unit_times **)a)->activating,
+                       (*(struct unit_times **)b)->activating);
 }
 
-static void free_unit_times(struct unit_times *t, unsigned n) {
-        struct unit_times *p;
+static void unit_times_freep(struct unit_times ***t) {
+        struct unit_times **p;
 
-        for (p = t; p < t + n; p++)
-                free(p->name);
+        if (!t || !*t)
```

Comment:
This will not deallocate an empty array, which seems wrong. I guess the check should be just `if (!t)`.

But I'm not sure if this level of pointer indirection is necessary: why not keep `struct unit_times *t` as it was, and just use fact that all allocated entries have `->name` set to determine the last one? I think this will make the code much easier to understand, triple pointers are  … tricky.

Also, `DEFINE_TRIVIAL_CLEANUP_FUNC` is nice because it does one level of pointer dereferencing and the NULL check, so the cleanup function itself is easier to grok.

So maybe something like this:
```C
static void free_unit_times(struct unit_times *t) {
     for (p = t; p->name; p++)
         free(p->name);
     free(t);
}
```
?

Reply:
I actually thought of that, but noticed there's a code path that [sets `name` to NULL](https://github.com/systemd/systemd/blob/81f6d448ef3cfb6331a68eb51dd6e63845a2bb51/src/analyze/analyze.c#L656) in the middle of a loop:

```c
for (u = times; u < times + n; u++) {
double text_start, text_width;

if (u->activating < boot->userspace_time ||
    u->activating > boot->finish_time) {
        u->name = mfree(u->name);
        continue;
}
```

Perhaps add another field to the structure to indicate whether it's valid?

I'll work on something like that and send a separate PR for it.